components/brave_shields/browser/ad_block_filters_provider_manager.cc

```diff
@@ -0,0 +1,105 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_shields/browser/ad_block_filters_provider_manager.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "base/barrier_callback.h"
+#include "base/location.h"
+#include "base/logging.h"
+#include "base/threading/sequenced_task_runner_handle.h"
+
+namespace brave_shields {
+
+namespace {
+
+static void OnDATLoaded(
+    base::OnceCallback<void(DATFileDataBuffer)> collect_and_merge,
+    bool deserialize,
+    const DATFileDataBuffer& dat_buf) {
+  // This manager should never be used for a provider that returns a serialized
+  // DAT. The ability should be removed from the FiltersProvider API when
+  // possible.
+  CHECK(!deserialize);
+
+  std::move(collect_and_merge).Run(dat_buf);
+}
+
+}  // namespace
+
+// static
+AdBlockFiltersProviderManager* AdBlockFiltersProviderManager::GetInstance() {
+  return base::Singleton<AdBlockFiltersProviderManager>::get();
+}
+
+AdBlockFiltersProviderManager::AdBlockFiltersProviderManager() {}
+
+AdBlockFiltersProviderManager::~AdBlockFiltersProviderManager() = default;
+
+void AdBlockFiltersProviderManager::AddProvider(
+    AdBlockFiltersProvider* provider) {
+  auto it =
+      std::find(filters_providers_.begin(), filters_providers_.end(), provider);
+  CHECK(it == filters_providers_.end());
+  filters_providers_.push_back(provider);
+  provider->AddObserver(this);
+}
+
+void AdBlockFiltersProviderManager::RemoveProvider(
+    AdBlockFiltersProvider* provider) {
+  auto it =
+      std::find(filters_providers_.begin(), filters_providers_.end(), provider);
+  CHECK(it != filters_providers_.end());
+  (*it)->RemoveObserver(this);
+  filters_providers_.erase(it);
+  NotifyObservers();
+}
+
+void AdBlockFiltersProviderManager::OnChanged() {
+  NotifyObservers();
+}
+
+void AdBlockFiltersProviderManager::LoadDATBuffer(
+    base::OnceCallback<void(bool deserialize, const DATFileDataBuffer& dat_buf)>
+        cb) {
+  if (task_tracker_.HasTrackedTasks()) {
+    // There's already an in-progress load, cancel it.
+    task_tracker_.TryCancelAll();
+  }
+
+  const auto collect_and_merge = base::BarrierCallback<DATFileDataBuffer>(
+      filters_providers_.size(),
+      base::BindOnce(&AdBlockFiltersProviderManager::FinishCombinating,
+                     weak_factory_.GetWeakPtr(), std::move(cb)));
+  for (auto* provider : filters_providers_) {
+    task_tracker_.PostTask(
+        base::SequencedTaskRunnerHandle::Get().get(), FROM_HERE,
+        base::BindOnce(
+            &AdBlockFiltersProvider::LoadDAT, base::Unretained(provider),
```

Comment:
is it a safe `Unretained`? What would happen if someone does `RemoveProvider` while the task is in progress